4.4.0+fio imx

ldts opened the pull request:
Rebase OP-TEE imx from lf-6.6.36_2.1.0 to  lf-6.6.52-2.2.0 

<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/en/latest/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

mike-scott commented:
@ldts I think this should need a CONFIG to specifically be enabled so that this doesn't end up in production devices.

ldts commented:
> @ldts I think this should need a CONFIG to specifically be enabled so that this doesn't end up in production devices.

I did code it to avoid a special config. It will have no impact on working production devices.

mike-scott commented:
I more

> > @ldts I think this should need a CONFIG to specifically be enabled so that this doesn't end up in production devices.
> 
> I did code it to avoid a special config. It will have no impact on working production devices.

I'm nervous that devices could end up being locked with the testing key and then treated as secure in production.  And from a usability stand-point users would never know that this was in fact a security liability (we've now published the test key in our git repos).

ldts commented:
um ok, but that should never happen..... also we have only published the SoC test key - not the device derived key. but anyway, ok will do a config. We will have to carry a special build for those customer boards that were messed up....will that be a branch in the factory? maybe tracking devel with the extra config?

ricardosalveti commented on core/arch/arm/plat-imx/drivers/imx_huk.c:
```diff
@@ -5,16 +5,46 @@
+         0xc2, 0x0c, 0x77, 0xec, 0xad, 0x89, 0xdc, 0x96,
+	 0xb7, 0x9f, 0xc8, 0xf7, 0xda, 0xab, 0x97, 0xb4,
+         0x2a, 0xe8, 0xdf, 0x98, 0x3d, 0x74, 0x1c, 0x34,
+	 0xac, 0xa8, 0x63, 0xca, 0xeb, 0x5f, 0xde, 0xcd,
```
Seems there is a mix of tabs and spaces here.

ricardosalveti commented on core/arch/arm/plat-imx/drivers/imx_huk.c:
```diff
@@ -5,16 +5,46 @@
 
+#if defined(CFG_MX7ULP)
+static uint8_t caam_use_test_hw_key;
+static uint8_t hw_test_key[] = {
```
Can you also add a comment here saying that you extracted this from a running device?

ricardosalveti commented on core/arch/arm/plat-imx/drivers/imx_huk.c:
```diff
@@ -5,16 +5,46 @@
 			return ret;
 		mkvb_retrieved = true;
 	}
+
```
No need for this extra line here.

ricardosalveti commented on core/tee/tee_rpmb_fs.c:
```diff
@@ -1096,9 +1097,9 @@ static TEE_Result tee_rpmb_init(uint16_t dev_id)
+retry_key:
 	if (!rpmb_ctx->key_derived) {
 		DMSG("RPMB INIT: Deriving key");
-
```
No need to remove this extra line.

ricardosalveti commented on core/arch/arm/plat-imx/drivers/imx_huk.c:
```diff
@@ -5,16 +5,46 @@
+
+TEE_Result tee_otp_enable_test_hw_unique_key(void)
+{
+	if (!caam_use_test_hw_key) {
```
Are we initializing this variable somewhere?

ldts replied:
is a global, it is initialized to 0

ricardosalveti replied:
Indeed.

ldts commented:
@ricardosalveti @mike-scott please have a look when you can. I enabled the config in the override for both - mfgtools and release
it is what is on the projects /lmp/builds/285/  experimental

ricardosalveti commented on core/arch/arm/plat-imx/drivers/imx_huk.c:
```diff
@@ -5,10 +5,46 @@
 	int ret = TEE_ERROR_SECURITY;
 
+#if defined(CFG_GET_ALTERNATIVE_HUK)
+	if (caam_use_test_hw_key) {
```
Can we guarantee that this function will not be called from another part of optee at the same time we set this to true later on?

ricardosalveti replied:
I guess I need to look back at optee to remember how often this function is actually called, as I don't remember it was called only once, with the output stored in memory.

ldts replied:
it is called from several places and and it will always default to the CAAM provided huk. Unless someone requests a _retry_ via calling the new API and then this function again ...in which case  it proceeds with the alternative...this works under the assumption there is only one thread of execution (so when during retry we enable the alternative, no other thread is going to request the huk)....

ldts replied:
since this is only supposed to be enabled on a few broken boards I didnt want to extend the list of parameters of the functions in the code path for maintainability reasons (hence the choice of making two calls leaving the default path unmodified)....we could do that and add an extra param on all the involved functions... just merging/upgrading op-tee might be a bit harder

ldts commented:
will we merge this ?it also need the overrides

vanmaegima committed:
[FIO internal] core: imx: remove HUK weak stub implementation

There's no use case for Foundries where we want this stub implementation
for HUK in most i.MX devices, so don't define this weak function for
i.MX6/7/8M/8MP to cause a build break.

CAAM driver for i.MX6UL is removed from OP-TEE. Now noone provides
the function tee_otp_get_hw_unique_key which is used in huk_subkey.c.
Add an exception for imx6ulevk in using a weak implementation of
tee_otp_get_hw_unique_key.

Signed-off-by: Vanessa Maegima <vanessa.maegima@foundries.io>
(cherry picked from commit f887178544984c87605963276bedabc08dc97bb8)
Co-developed--by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>
Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>

ricardosalveti committed:
[FIO extras] ta: pkcs11: change UUID to avoid conflict with SKS TA

Upstream pkcs11 TA shares the same UUID as previously used by SKS, but
they are not compatible with each other from the storage perspective, so
change UUID to avoid conflict and facilitate the transition by the user
by allowing both TAs to be installed at the same time.

Signed-off-by: Ricardo Salveti <ricardo@foundries.io>
(cherry picked from commit e2ad56f0ab28083051df98fbe959ff522fa6daaa)

ldts committed:
[FIO extras] drivers: crypto: se050: import ecc/rsa keys

Signed-off-by: Jorge Ramirez-Ortiz <jorge@foundries.io>
Signed-off-by: Ricardo Salveti <ricardo@foundries.io>
(cherry picked from commit 73c9ad85a55894661bbd6b37316aa4b3481b6052)

ldts committed:
[FIO extras] ta: pkcs11: secure element: import pre-allocated keys

Signed-off-by: Jorge Ramirez-Ortiz <jorge@foundries.io>
Signed-off-by: Ricardo Salveti <ricardo@foundries.io>
(cherry picked from commit 19f0d4a0ee53a49387ac2714a20d84e953b9d9fe)

ricardosalveti committed:
[FIO extras] ta: pkcs11: use deviceID directly as token description

To make it compatible with the way used by SKS, drop the "OP-TEE PKCS11
TA - TEE UUID" string from the slot information and let it be set
directly to the deviceID (UUID).

We can drop this change once we convert all our userspace to be
compatible with both formats.

Signed-off-by: Ricardo Salveti <ricardo@foundries.io>
(cherry picked from commit 265e4aef62ac1efecea6b75865eba769168b2b73)

ldts committed:
[FIO internal]: se050: scp03: prohibit derived keys without RPMB ready.

Prevents SCP03 rotation from using the derived keys unless
RPMB is available.

This is because RPMB available means a stable/inmutable HUK.

This commit must be used particularly when
CFG_CORE_SE05X_SCP03_PROVISION_ON_INIT has been enabled.

Signed-off-by: Jorge Ramirez-Ortiz <jorge@foundries.io>
(cherry picked from commit bd6269660d198d1887fd4f0038d7430ea840c454)
